File: java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/service/NotificationService.java
package io.hcxprotocol.service;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.typesafe.config.Config;
import io.hcxprotocol.dto.NotificationRequest;
import io.hcxprotocol.exception.ClientException;
import io.hcxprotocol.exception.ErrorCodes;
import io.hcxprotocol.utils.*;
import org.apache.commons.text.StringSubstitutor;

import java.util.*;

import static io.hcxprotocol.utils.Constants.*;

public class NotificationService {

    public static void validateNotificationRequest(NotificationRequest notificationRequest) throws ClientException {
        if (notificationRequest.getTopicCode().isEmpty()) {
            throw new ClientException("Topic code cannot be empty");
        } else if (notificationRequest.getMessage().isEmpty() && notificationRequest.getTemplateParams().isEmpty()) {
            throw new ClientException("Either the message or the template parameters are mandatory.");
        } else if (notificationRequest.getRecipients().isEmpty()) {
            throw new ClientException("Recipients cannot be empty");
        } else if (notificationRequest.getRecipientType().isEmpty()) {
            throw new ClientException("Recipient type cannot be empty");
        }
    }

    public static String getMessage(NotificationRequest notificationRequest,Map<String, Object> output, String message) throws Exception {
        Map<String, Object> searchRequest = Map.of("filters", new HashMap<>());
        if (message.isEmpty()) {
            Map<String, Object> responseMap = new HashMap<>();
            Utils.initializeHCXCall(JSONUtils.serialize(searchRequest), Operations.NOTIFICATION_LIST, responseMap, notificationRequest.getConfig());
            output.putAll(responseMap);
            if (output.containsKey(Constants.ERROR) || output.containsKey(ErrorCodes.ERR_DOMAIN_PROCESSING.toString()))
                throw new ClientException("Error while resolving the message template");
            List<Map<String, Object>> notificationsList = ((List<Map<String, Object>>) ((Map<String, Object>) responseMap.get("responseObj")).get("notifications"));
            Map<String, Object> notification = getNotification(notificationsList, notificationRequest.getTopicCode());
            message = resolveTemplate(notification, notificationRequest.getTemplateParams());
        }
        return message;
    }

    public static Map<String, Object> getNotification(List<Map<String, Object>> notificationList, String code) {
        Map<String, Object> notification = new HashMap<>();
        Optional<Map<String, Object>> result = notificationList.stream().filter(obj -> obj.get(TOPIC_CODE).equals(code)).findFirst();
        if (result.isPresent()) notification = result.get();
        return notification;
    }

    public static String getPrivateKey(Config config) {
        String privateKey = config.getString("signingPrivateKey");
        privateKey = privateKey
                .replace("-----BEGIN PRIVATE KEY-----", "")
                .replace("-----END PRIVATE KEY-----", "")
                .replaceAll("\\s+", "");
        return privateKey;
    }

    public static String resolveTemplate(Map<String, Object> notification, Map<String, String> nData) throws JsonProcessingException {
        StringSubstitutor sub = new StringSubstitutor(nData);
        return sub.replace((JSONUtils.deserialize((String) notification.get(Constants.TEMPLATE), Map.class)).get(Constants.MESSAGE));
    }

    public static Map<String, Object> getJWSRequestHeader(NotificationRequest notificationRequest) {
        Map<String, Object> headersMap = new HashMap<>();
        headersMap.put("x-hcx-correlation_id", notificationRequest.getCorrelationId().isEmpty() ? UUID.randomUUID() : notificationRequest.getCorrelationId());
        headersMap.put("sender_code", notificationRequest.getConfig().getString(Constants.PARTICIPANT_CODE));
        headersMap.put(Constants.TIMESTAMP, System.currentTimeMillis());
        headersMap.put(RECIPIENT_TYPE, notificationRequest.getRecipientType());
        headersMap.put(RECIPIENTS, notificationRequest.getRecipients());
        Map<String, Object> headers = new HashMap<>();
        headers.put("alg", "RS256");
        headers.put(NOTIFICATION_HEADERS, headersMap);
        return headers;
    }

    public static Map<String, Object> getJWSRequestPayload(NotificationRequest notificationRequest, Map<String, Object> output, String message) throws Exception {
        Map<String, Object> payload = new HashMap<>();
        payload.put(TOPIC_CODE, notificationRequest.getTopicCode());
        message = NotificationService.getMessage(notificationRequest, output,message);
        payload.put(MESSAGE, message);
        return payload;
    }

    public static Map<String,Object> createNotificationRequest(NotificationRequest notificationRequest,Map<String,Object> output,String message) throws Exception {
        Map<String, Object> headers = NotificationService.getJWSRequestHeader(notificationRequest);
        Map<String,Object> payload = NotificationService.getJWSRequestPayload(notificationRequest,output,message);
        String jwsToken = JWTUtils.generateJWS(headers, payload, NotificationService.getPrivateKey(notificationRequest.getConfig()));
        Map<String, Object> requestBody = new HashMap<>();
        requestBody.put(PAYLOAD, jwsToken);
        return requestBody;
    }
}

File: dot-net/hcx-integrator-sdk/Io.HcxProtocol/Service/NotificationService.cs
using Io.HcxProtocol.Dto;
using Io.HcxProtocol.Exceptions;
using Io.HcxProtocol.Helper;
using Io.HcxProtocol.Init;
using Io.HcxProtocol.Utils;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Io.HcxProtocol.Service
{
    public class NotificationService
    {
        private NotificationService() { }

        public static void ValidateNotificationRequest(NotificationRequest notificationRequest)
        {
            if (string.IsNullOrEmpty(notificationRequest.GetTopicCode()))
            {
                throw new ClientException("Topic code cannot be empty.");
            }
            else if (string.IsNullOrEmpty(notificationRequest.GetMessage()) && notificationRequest.GetTemplateParams().Count == 0)
            {
                throw new ClientException("Either the message or the template parameters are mandatory.");
            }
            else if (notificationRequest.GetRecipients().Count == 0)
            {
                throw new ClientException("Recipients cannot be empty.");
            }
            else if (string.IsNullOrEmpty(notificationRequest.GetRecipientType()))
            {
                throw new ClientException("Recipient type cannot be empty.");
            }
        }

        public static Dictionary<string, object> CreateNotificationRequest(NotificationRequest notificationRequest, Dictionary<string, object> output, string message)
        {
            Dictionary<string, object> headers = GetJWSRequestHeader(notificationRequest);
            Dictionary<string, object> payload = GetJWSRequestPayload(notificationRequest, output, message);
            string privatekey = GetPrivateKey(notificationRequest.GetConfig());
            string jwsToken = JWSUtils.Generate(headers, payload, privatekey);
            Dictionary<string, object> requestBody = new Dictionary<string, object>();
            requestBody.Add("payload", jwsToken);
            return requestBody;
        }

        private static Dictionary<string, object> GetJWSRequestHeader(NotificationRequest notificationRequest)
        {
            var guid = Guid.NewGuid().ToString();
            Dictionary<string, object> headersDictionary = new Dictionary<string, object>();
            headersDictionary.Add("x-hcx-correlation_id", string.IsNullOrEmpty(notificationRequest.GetCorrelationId()) ? guid : notificationRequest.GetCorrelationId());
            headersDictionary.Add("sender_code", notificationRequest.GetConfig().ParticipantCode);
            headersDictionary.Add(Constants.TIMESTAMP, GetCurrentMilli());
            headersDictionary.Add(Constants.RECIPIENT_TYPE, notificationRequest.GetRecipientType());
            headersDictionary.Add(Constants.RECIPIENTS, notificationRequest.GetRecipients());
            Dictionary<string, object> headers = new Dictionary<string, object>();
            headers.Add("alg", "RS256");
            headers.Add(Constants.NOTIFICATION_HEADERS, headersDictionary);
            return headers;
        }

        private static Dictionary<string, object> GetJWSRequestPayload(NotificationRequest notificationRequest, Dictionary<string, object> output, string message)
        {
            Dictionary<string, object> payload = new Dictionary<string, object>();
            payload.Add(Constants.TOPIC_CODE, notificationRequest.GetTopicCode());
            message = GetMessage(notificationRequest, output, message);
            payload.Add(Constants.MESSAGE, message);
            return payload;
        }

        private static string GetMessage(NotificationRequest notificationRequest, Dictionary<string, object> output, string message)
        {
            Dictionary<string, object> searchRequest = new Dictionary<string, object>
            {
                {"filters", new Dictionary<string, object>()}
            };
            if (notificationRequest.GetTemplateParams().Count != 0)
            {
                HcxUtils.InitializeHCXCall(JSONUtils.Serialize(searchRequest), Operations.NOTIFICATION_LIST, output, notificationRequest.GetConfig());
                if (output.ContainsKey(Constants.ERROR) || output.ContainsKey(ErrorCodes.ERR_DOMAIN_PROCESSING.ToString()))
                {
                    throw new ClientException("Error while resolving the message template.");
                }
                if (output.Count > 0)
                {
                    var notificationsList = output["responseObj"];
                    Dictionary<object, object> dictio = (Dictionary<object, object>)notificationsList;
                    Dictionary<string, object> dictiotwo = dictio.ToDictionary(x => x.Key.ToString(), x => (object)x.Value);

                    List<Dictionary<string, object>> notificationlist = new List<Dictionary<string, object>>();
                    notificationlist.Add(dictiotwo);
                    var topiccode = notificationRequest.GetTopicCode();
                    Dictionary<string, object> notification = GetNotification(notificationlist, topiccode);
                    message = ResolveTemplate(notification, notificationRequest.GetTemplateParams());
                }
            }
            return message;
        }

        private static string ResolveTemplate(Dictionary<string, object> notification, Dictionary<string, string> nData)
        {
            var template = JsonConvert.SerializeObject(notification[Constants.TEMPLATE]);
            var versionnumber = nData["version_code"];
            var participantname = nData["participant_name"];
            template = template.Replace("${participant_name}", participantname);
            template = template.Replace("${version_code}", versionnumber);
            return template.ToString();
        }

        private static Dictionary<string, object> GetNotification(List<Dictionary<string, object>> notificationList, string code)
        {
            var list = notificationList[0];
            var finalist = list["notifications"];
            var d = JsonConvert.SerializeObject(finalist.ToString());

            var type = finalist.GetType();
            Dictionary<string, object> x = notificationList.Where(x1 => x1.ContainsKey("notifications")).FirstOrDefault();
            var j = x.Where(obj1 => obj1.Key.Equals("notifications")).FirstOrDefault().Value;
            var lisstring = j.ToString();
            var data = JsonConvert.DeserializeObject<IEnumerable<IDictionary<string, object>>>(lisstring);
            var notification = data.Where(mo => mo["topic_code"].Equals(code)).First();
            var dictioobj = notification.ToDictionary(co => co.Key.ToString(), co => (object)co.Value);
            if (dictioobj.Count > 1)
            {
                return dictioobj;
            }
            else
            {
                throw new ClientException("Topic code is not found in the notification list: " + code);
            }
        }

        private static string GetPrivateKey(Config config)
        {
            return config.SigningPrivateKey;
        }

        public static double GetCurrentMilli()
        {
            DateTime Jan1970 = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
            TimeSpan javaSpan = DateTime.UtcNow - Jan1970;
            var x = decimal.Round(Convert.ToDecimal(javaSpan.TotalMilliseconds));
            return Convert.ToDouble(x);
        }
    }
}
